multiversx_sdk_core/transaction_factories/smart_contract_factory.py

```diff
@@ -0,0 +1,192 @@
+               value: Optional[ITransactionValue] = None,
+               gas_price: Optional[IGasPrice] = None,
+               is_upgradeable: bool = True,
+               is_readable: bool = True,
+               is_payable: bool = False,
+               is_payable_by_sc: bool = False) -> Transaction:
+        bytecode = read_binary_file(bytecode_path)
+        metadata = CodeMetadata(is_upgradeable, is_readable, is_payable, is_payable_by_sc)
+        arguments = arguments or []
+
+        parts = [
+            arg_to_string(bytecode),
+            arg_to_string(VM_TYPE_WASM_VM),
+            metadata.to_hex()
+        ]
+
+        for arg in arguments:
+            parts.append(arg_to_string(arg))
+
+        transaction = self.create_transaction(
+            sender=deployer,
+            receiver=Address.from_bech32(CONTRACT_DEPLOY_ADDRESS),
+            data_parts=parts,
+            gas_limit=gas_limit,
+            gas_price=gas_price,
+            nonce=nonce,
+            value=value
+        )
+
+        return transaction
+
+    def execute(self,
+                sender: IAddress,
+                contract_address: IAddress,
+                function: str,
+                gas_limit: IGasLimit,
+                arguments: Optional[List[Any]] = None,
+                nonce: Optional[INonce] = None
+                ) -> Transaction:
+        arguments = arguments or []
+
+        parts = [function]
+
+        for arg in arguments:
+            parts.append(arg_to_string(arg))
+
+        trasaction = self.create_transaction(
+            sender=sender,
+            receiver=contract_address,
+            data_parts=parts,
+            gas_limit=gas_limit,
+            nonce=nonce
+        )
+
+        return trasaction
+
+    def upgrade(self,
+                sender: IAddress,
+                contract: IAddress,
+                bytecode_path: Path,
+                gas_limit: IGasLimit,
+                arguments: Optional[List[Any]] = None,
+                nonce: Optional[INonce] = None,
+                is_upgradeable: bool = True,
+                is_readable: bool = True,
+                is_payable: bool = False,
+                is_payable_by_sc: bool = False
+                ) -> Transaction:
+        arguments = arguments or []
+        bytecode = read_binary_file(bytecode_path)
+        metadata = CodeMetadata(is_upgradeable, is_readable, is_payable, is_payable_by_sc)
+
+        parts = [
+            "upgradeContract",
+            arg_to_string(bytecode),
+            metadata.to_hex()
+        ]
+
+        for arg in arguments:
+            parts.append(arg_to_string(arg))
+
+        transaction = self.create_transaction(
+            sender=sender,
+            receiver=contract,
+            data_parts=parts,
+            gas_limit=gas_limit,
+            nonce=nonce
+        )
+
+        return transaction
+
+    def compute_contract_address(self, deployer: IAddress, nonce: INonce) -> Address:
```

Comment:
We have the existing implementation here:

https://github.com/multiversx/mx-sdk-py-core/blob/main/multiversx_sdk_core/address.py#L142